File: packages/python/qsu/crypto/decrypt.py
import base64

from cryptography.hazmat.primitives import padding

from ._cipher import (
	AEAD_MODES,
	BLOCK_SIZE_BITS,
	PADDED_MODES,
	buildCipher,
	parseAlgorithm,
)


def decrypt(
	str: str,
	secret: str,
	algorithm: str = 'aes-256-cbc',
	toBase64: bool = False,
) -> str:
	if not str or len(str) < 1:
		return ''

	key, modeName = parseAlgorithm(algorithm, secret)
	isAead = modeName in AEAD_MODES
	arrStr = str.split(':')

	if len(arrStr) < (3 if isAead else 2):
		raise ValueError(
			'`str` must be in the `iv:authTag:encrypted` format returned by `encrypt`.'
			if isAead
			else '`str` must be in the `iv:encrypted` format returned by `encrypt`.'
		)

	def decode(value):
		return base64.b64decode(value) if toBase64 else bytes.fromhex(value)

	iv = decode(arrStr.pop(0))
	tag = decode(arrStr.pop(0)) if isAead else None
	enc = decode(':'.join(arrStr))

	decryptor = buildCipher(key, modeName, iv, tag).decryptor()
	decrypted = decryptor.update(enc) + decryptor.finalize()

	if modeName in PADDED_MODES:
		# Let PKCS7 validate the padding. Reading the last byte and slicing by it accepted
		# any garbage, so decrypting with a wrong key returned '' instead of failing.
		unpadder = padding.PKCS7(BLOCK_SIZE_BITS).unpadder()
		decrypted = unpadder.update(decrypted) + unpadder.finalize()

	return decrypted.decode('utf-8')

File: packages/javascript/lib/node/crypto/decrypt.ts
import { createDecipheriv } from 'node:crypto';
import { AEAD_ALGORITHM } from './encrypt.js';

export function decrypt(
	str: string,
	secret: string,
	algorithm = 'aes-256-cbc',
	toBase64 = false
): string {
	if (!str || str.length < 1) {
		return '';
	}

	const encoding: BufferEncoding = toBase64 ? 'base64' : 'hex';
	const arrStr: string[] = str.split(':');
	const isAead = AEAD_ALGORITHM.test(algorithm);

	if (arrStr.length < (isAead ? 3 : 2)) {
		throw new Error(
			isAead
				? '`str` must be in the `iv:authTag:encrypted` format returned by `encrypt`.'
				: '`str` must be in the `iv:encrypted` format returned by `encrypt`.'
		);
	}

	const iv = Buffer.from(arrStr.shift() as string, encoding);
	const authTag = isAead ? Buffer.from(arrStr.shift() as string, encoding) : null;
	const decipher = createDecipheriv(algorithm, secret, iv);

	if (authTag) {
		(decipher as unknown as { setAuthTag: (tag: Buffer) => void }).setAuthTag(authTag);
	}

	let decrypted = decipher.update(Buffer.from(arrStr.join(':'), encoding));

	decrypted = Buffer.concat([decrypted, decipher.final()]);

	return decrypted.toString();
}
